zebra/zserv.c

```diff
@@ -2754,157 +2691,206 @@ static void zserv_write_incoming(struct stream *orig, uint16_t command)
 }
 #endif
 
+static int zserv_process_messages(struct thread *thread)
+{
+	struct zserv *client = THREAD_ARG(thread);
+	struct zebra_vrf *zvrf;
+	struct zmsghdr hdr;
+	struct stream *msg;
+	bool hdrvalid;
+
+	do {
+		msg = stream_fifo_pop(client->ibuf_fifo);
+
+		/* break if out of messages */
+		if (!msg)
+			continue;
+
+		/* read & check header */
+		hdrvalid = zapi_parse_header(msg, &hdr);
+		if (!hdrvalid && IS_ZEBRA_DEBUG_PACKET && IS_ZEBRA_DEBUG_RECV) {
+			const char *emsg = "Message has corrupt header";
+			zserv_log_message(emsg, msg, NULL);
+		}
+		if (!hdrvalid)
+			continue;
+
+		/* lookup vrf */
+		zvrf = zebra_vrf_lookup_by_id(hdr.vrf_id);
+		if (!zvrf && IS_ZEBRA_DEBUG_PACKET && IS_ZEBRA_DEBUG_RECV) {
+			const char *emsg = "Message specifies unknown VRF";
+			zserv_log_message(emsg, msg, &hdr);
+		}
+		if (!zvrf)
+			continue;
+
+		/* process commands */
+		zserv_handle_commands(client, &hdr, msg, zvrf);
+
+	} while (msg);
```

Comment:
In `zebra_client_read()` we read up to `zebrad.packets_to_process` packets at time. Wouldn't it be a good idea to have a limit like that here as well? I think it will be necessary especially when the ZAPI I/O is segregated into another thread reading packets non-stop. And instead of `zebrad.packets_to_process` we could try using `thread_should_yield()` but I'm not sure if it would work.

Reply:
You are correct, I have code for this on my multithreaded ZAPI branch. I can pull it into this PR if you think it belongs here.

Reply:
Ok, adding a limit later together with the MT ZAPI code sounds reasonable to me.